File: backend/src/core/monitoring.py
"""Monitoring and error tracking setup."""
import os
from typing import Optional
from src.core.logging import get_logger

logger = get_logger(__name__)


class ErrorTracker:
    """
    Error tracking integration (Sentry, etc.).

    This class provides a unified interface for error tracking services.
    Currently configured for Sentry, but can be extended for other services.
    """

    def __init__(self):
        self.enabled = False
        self.sentry_dsn = os.getenv("SENTRY_DSN")

        if self.sentry_dsn:
            self._init_sentry()

    def _init_sentry(self):
        """Initialize Sentry error tracking."""
        try:
            import sentry_sdk
            from sentry_sdk.integrations.fastapi import FastApiIntegration
            from sentry_sdk.integrations.starlette import StarletteIntegration

            sentry_sdk.init(
                dsn=self.sentry_dsn,
                # Set traces_sample_rate to 1.0 to capture 100% of transactions for performance monitoring.
                # Adjust this value in production to reduce costs.
                traces_sample_rate=float(os.getenv("SENTRY_TRACES_SAMPLE_RATE", "0.1")),
                # Set profiles_sample_rate to profile 10% of sampled transactions.
                profiles_sample_rate=float(os.getenv("SENTRY_PROFILES_SAMPLE_RATE", "0.1")),
                environment=os.getenv("ENVIRONMENT", "development"),
                integrations=[
                    StarletteIntegration(),
                    FastApiIntegration(),
                ],
            )

            self.enabled = True
            logger.info("Sentry error tracking initialized")

        except ImportError:
            logger.warning(
                "Sentry SDK not installed. Install with: uv add sentry-sdk[fastapi]"
            )
        except Exception as e:
            logger.error(f"Failed to initialize Sentry: {e}")

    def capture_exception(
        self,
        exception: Exception,
        context: Optional[dict] = None
    ):
        """
        Capture an exception for error tracking.

        Args:
            exception: The exception to capture
            context: Additional context to attach to the error
        """
        if not self.enabled:
            logger.error(
                f"Exception occurred (error tracking disabled): {exception}",
                extra=context or {},
                exc_info=True
            )
            return

        try:
            import sentry_sdk

            if context:
                with sentry_sdk.push_scope() as scope:
                    for key, value in context.items():
                        scope.set_extra(key, value)
                    sentry_sdk.capture_exception(exception)
            else:
                sentry_sdk.capture_exception(exception)

        except Exception as e:
            logger.error(f"Failed to capture exception in Sentry: {e}")

    def capture_message(
        self,
        message: str,
        level: str = "info",
        context: Optional[dict] = None
    ):
        """
        Capture a message for tracking.

        Args:
            message: The message to capture
            level: Severity level (debug, info, warning, error, fatal)
            context: Additional context
        """
        if not self.enabled:
            getattr(logger, level, logger.info)(message, extra=context or {})
            return

        try:
            import sentry_sdk

            if context:
                with sentry_sdk.push_scope() as scope:
                    for key, value in context.items():
                        scope.set_extra(key, value)
                    sentry_sdk.capture_message(message, level=level)
            else:
                sentry_sdk.capture_message(message, level=level)

        except Exception as e:
            logger.error(f"Failed to capture message in Sentry: {e}")


class PerformanceMonitor:
    """
    Performance monitoring for API endpoints.

    Tracks response times, throughput, and other performance metrics.
    """

    def __init__(self):
        self.enabled = os.getenv("ENABLE_PERFORMANCE_MONITORING", "false").lower() == "true"
        self.metrics = {
            "requests": 0,
            "errors": 0,
            "total_duration_ms": 0,
        }

    def record_request(self, duration_ms: float, status_code: int):
        """
        Record a request for performance tracking.

        Args:
            duration_ms: Request duration in milliseconds
            status_code: HTTP status code
        """
        if not self.enabled:
            return

        self.metrics["requests"] += 1
        self.metrics["total_duration_ms"] += duration_ms

        if status_code >= 400:
            self.metrics["errors"] += 1

    def get_stats(self) -> dict:
        """Get current performance statistics."""
        total_requests = self.metrics["requests"]

        return {
            "total_requests": total_requests,
            "total_errors": self.metrics["errors"],
            "error_rate": self.metrics["errors"] / total_requests if total_requests > 0 else 0,
            "avg_response_time_ms": (
                self.metrics["total_duration_ms"] / total_requests
                if total_requests > 0
                else 0
            ),
        }

    def reset_stats(self):
        """Reset performance statistics."""
        self.metrics = {
            "requests": 0,
            "errors": 0,
            "total_duration_ms": 0,
        }


# Global instances
error_tracker = ErrorTracker()
performance_monitor = PerformanceMonitor()


def get_error_tracker() -> ErrorTracker:
    """Get the global error tracker instance."""
    return error_tracker


def get_performance_monitor() -> PerformanceMonitor:
    """Get the global performance monitor instance."""
    return performance_monitor

File: frontend/lib/monitoring.ts
/**
 * Frontend error tracking and monitoring
 *
 * Provides error tracking integration (Sentry, etc.) for the frontend.
 * Can be configured via environment variables.
 */

interface ErrorContext {
  [key: string]: any;
}

class ErrorTracker {
  private enabled: boolean = false;
  private sentryDsn: string | undefined;

  constructor() {
    this.sentryDsn = process.env.NEXT_PUBLIC_SENTRY_DSN;

    if (this.sentryDsn && typeof window !== 'undefined') {
      this.initSentry();
    }
  }

  private initSentry() {
    try {
      // Dynamic import to avoid SSR issues
      import('@sentry/nextjs').then((Sentry) => {
        Sentry.init({
          dsn: this.sentryDsn,
          environment: process.env.NEXT_PUBLIC_ENVIRONMENT || 'development',
          tracesSampleRate: parseFloat(process.env.NEXT_PUBLIC_SENTRY_TRACES_SAMPLE_RATE || '0.1'),
          // Session Replay
          replaysSessionSampleRate: 0.1,
          replaysOnErrorSampleRate: 1.0,
        });

        this.enabled = true;
        console.log('[Monitoring] Sentry initialized');
      }).catch((error) => {
        console.warn('[Monitoring] Sentry not available:', error.message);
      });
    } catch (error) {
      console.error('[Monitoring] Failed to initialize Sentry:', error);
    }
  }

  /**
   * Capture an exception
   */
  captureException(error: Error, context?: ErrorContext) {
    if (!this.enabled) {
      console.error('[Error]', error, context);
      return;
    }

    import('@sentry/nextjs').then((Sentry) => {
      if (context) {
        Sentry.withScope((scope) => {
          Object.entries(context).forEach(([key, value]) => {
            scope.setExtra(key, value);
          });
          Sentry.captureException(error);
        });
      } else {
        Sentry.captureException(error);
      }
    });
  }

  /**
   * Capture a message
   */
  captureMessage(message: string, level: 'info' | 'warning' | 'error' = 'info', context?: ErrorContext) {
    if (!this.enabled) {
      console[level === 'warning' ? 'warn' : level]('[Message]', message, context);
      return;
    }

    import('@sentry/nextjs').then((Sentry) => {
      if (context) {
        Sentry.withScope((scope) => {
          Object.entries(context).forEach(([key, value]) => {
            scope.setExtra(key, value);
          });
          Sentry.captureMessage(message, level);
        });
      } else {
        Sentry.captureMessage(message, level);
      }
    });
  }

  /**
   * Set user context
   */
  setUser(user: { id: string; email?: string; name?: string } | null) {
    if (!this.enabled) return;

    import('@sentry/nextjs').then((Sentry) => {
      Sentry.setUser(user);
    });
  }

  /**
   * Add breadcrumb for debugging
   */
  addBreadcrumb(message: string, category: string, data?: any) {
    if (!this.enabled) return;

    import('@sentry/nextjs').then((Sentry) => {
      Sentry.addBreadcrumb({
        message,
        category,
        data,
        level: 'info',
      });
    });
  }
}

// Global instance
export const errorTracker = new ErrorTracker();

/**
 * Log an error with context
 */
export function logError(error: Error, context?: ErrorContext) {
  errorTracker.captureException(error, context);
}

/**
 * Log a message
 */
export function logMessage(message: string, level: 'info' | 'warning' | 'error' = 'info', context?: ErrorContext) {
  errorTracker.captureMessage(message, level, context);
}

/**
 * Set user context for error tracking
 */
export function setUserContext(user: { id: string; email?: string; name?: string } | null) {
  errorTracker.setUser(user);
}

/**
 * Add breadcrumb for debugging
 */
export function addBreadcrumb(message: string, category: string = 'navigation', data?: any) {
  errorTracker.addBreadcrumb(message, category, data);
}

/**
 * Performance monitoring
 */
export class PerformanceMonitor {
  private marks: Map<string, number> = new Map();

  /**
   * Start measuring an operation
   */
  start(operationName: string) {
    this.marks.set(operationName, performance.now());
  }

  /**
   * End measuring and log the duration
   */
  end(operationName: string, context?: ErrorContext) {
    const startTime = this.marks.get(operationName);
    if (!startTime) {
      console.warn(`[Performance] No start mark found for: ${operationName}`);
      return;
    }

    const duration = performance.now() - startTime;
    this.marks.delete(operationName);

    // Log slow operations (> 1 second)
    if (duration > 1000) {
      logMessage(
        `Slow operation detected: ${operationName}`,
        'warning',
        { duration_ms: duration, ...context }
      );
    }

    return duration;
  }
}

export const performanceMonitor = new PerformanceMonitor();
